packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,354 @@
+
+  var pt = resp.message.publishTime;
+  var milliseconds = parseInt(pt.nanos, 10) / 1e6;
+  var data = resp.message.data;
+
+  return {
+    connectionId: connectionId,
+    ackId: resp.ackId,
+    id: resp.message.messageId,
+    attributes: resp.message.attributes,
+    publishTime: new Date(parseInt(pt.seconds, 10) * 1000 + milliseconds),
+    received: Date.now(),
+    // using get here to prevent user from overwriting data
+    get data() {
+      return data;
+    },
+    ack: function() {
+      self.subscription.ack_(this);
+    },
+    nack: function() {
+      self.subscription.nack_(this);
+    }
+  };
+};
+
+/**
+ * Creates specified number of connections and puts pool in open state.
+ */
+ConnectionPool.prototype.open = function() {
+  var existing = this.connections.size;
+  var max = this.settings.maxConnections;
+
+  for (; existing < max; existing++) {
+    this.queueConnection();
+  }
+
+  this.isOpen = true;
+};
+
+/**
+ * Pauses each of the connections, causing `message` events to stop firing.
+ */
+ConnectionPool.prototype.pause = function() {
+  this.isPaused = true;
+
+  this.connections.forEach(function(connection) {
+    connection.pause();
+  });
+};
+
+/**
+ * Queues a connection to be created. If any previous connections have failed,
+ * it will apply a back off based on the number of failures.
+ */
+ConnectionPool.prototype.queueConnection = function() {
+  var self = this;
+
+  var attempts = this.failedConnectionAttempts / this.settings.maxConnections;
+  var delay = (Math.pow(2, Math.ceil(attempts)) * 1000) +
+    (Math.floor(Math.random() * 1000));
+
+  this.queue = this.queue.then(function() {
+    return common.util.promisify(self.createConnection).call(self);
+  }).then(function() {
+    return new Promise(function(resolve) {
+      setTimeout(resolve, delay);
+    });
+  });
+};
+
+/**
+ * Calls resume on each connection, allowing `message` events to fire off again.
+ */
+ConnectionPool.prototype.resume = function() {
+  this.isPaused = false;
+
+  this.connections.forEach(function(connection) {
+    connection.resume();
+  });
+};
+
+/**
+ * Inspects a status object to determine whether or not we should try and
+ * reconnect.
+ *
+ * @param {object} status - The gRPC status object.
+ * @return {boolean}
+ */
+ConnectionPool.prototype.shouldReconnect = function(status) {
+  // If the pool was closed, we should definitely not reconnect
+  if (!this.isOpen) {
+    return false;
+  }
+
+  // We should check to see if the status code is a non-recoverable error
+  if (RETRY_CODES.indexOf(status.code) === -1) {
+    return false;
+  }
+
+  // deadline exceeded errors are tricky because gax applies a deadline no
+  // matter what, so it might not be an error at all, the stream could have just
+  // been closed. That being said we need to check if it is a deadline error
+  // and if it is only stop retrying if we've failed to make a connection
+  var elapsed = Date.now() - this.lastKnownConnection;
```

Comment:
I'm not certain this is right:

- this.lastKnownConnection tracks a time a connection was opened. But you want to make a judgement based on the time a connection closed (for example, in healthy operation, the last time a connection opened might be minutes, hours, or days ago)
- You should propagate the error if there are no active connections and have not been for some period of time. I don't see how this is accounted for here